### PR TITLE
feat: initial implementation of macOS sign in handling

### PR DESCRIPTION
**Pull request by Gustash**
## Changes

This PR enables proper support for custom protocols in macOS. Primarily, this fixes the signin by enabling macOS to handle the `qrc://` protocol in `MaximaBootstrap.app`.

For this, you first need to run `cargo install --git https://github.com/burtonageo/cargo-bundle` to get the latest version of cargo-bundle. Then you need to run `cargo bundle` inside the `maxima-bootstrap` directory. This will generate the `MaximaBootstrap.app` app bundle, which is required to handle the custom protocols.

## TODO

This can be done in a separate PR, but right now cargo-bundle does not support setting the `LSUIElement` property in the Info.plist of the generated app bundle. We need this property to tell the system that this app bundle has no UI, so it shows no signs of it in the Dock, Menu Bar, etc.

We'll probably need to fork cargo-bundle for this since it's developments seems quite stagnant.

### CODE REVIEW

File: maxima-bootstrap/Cargo.toml [hunk 1]
```diff
@@ -1,5 +1,6 @@
 [package]
 name = "maxima-bootstrap"
+description = "Maxima handler for custom EA protocols"
```
Comment (battledash) [inline maxima-bootstrap/Cargo.toml:3]: I feel this is a little too generic for what bootstrap does, it also serves as the installer for the windows service and as an intermediary process for launching games. I'm not entirely sure what something better and still concise would be though

File: maxima-bootstrap/src/main.rs [hunk 2]
```diff
@@ -103,9 +138,9 @@ async fn run(args: &Vec<String>) -> Result<()> {
...
     }
-    
-    if len > 2 {
-        let command = &args[1];
+
+    if len > 1 {
+        let command = &args[0];
         match command.as_str() {
             "launch" => {
                 let decoded = general_purpose::STANDARD.decode(&args[2])?;
```
Comment (battledash) [inline maxima-bootstrap/src/main.rs]: We're now reading `0` and `2` here, which will break game launching.

File: maxima-bootstrap/src/main.rs [hunk 3]
```diff
@@ -78,10 +103,10 @@ fn platform_launch(args: BootstrapLaunchArgs) -> Result<()> {
     Ok(())
 }
 
-async fn run(args: &Vec<String>) -> Result<()> {
+async fn run(args: &[String]) -> Result<()> {
     let len = args.len();
-    if len == 2 {
-        let arg = &args[1];
+    if len == 1 {
+        let arg = &args[0];
```
Comment (battledash) [inline maxima-bootstrap/src/main.rs:112]: What's the reason for this change? The first entry in args is usually the executable path, so this theoretically would break the handlers we have that use this value. If it's different on macos we'll need a special case for that, but on win/unix the first entry is always the executable

File: maxima-lib/src/util/registry.rs [hunk 4]
```diff
@@ -165,21 +150,40 @@ fn register_custom_protocol(protocol: &str, name: &str, executable: &str) -> Res
...
+#[cfg(target_os = "macos")]
+pub fn set_up_registry() -> Result<()> {
+    use std::process::Command;
+
+    use log::warn;
+
+    let bin = bootstrap_path()
+        .join("Contents")
+        .join("MacOS")
+        .join("maxima-bootstrap");
```
Comment (battledash) [inline maxima-lib/src/util/registry.rs]: `bootstrap_path` should return the bootstrap executable, so there shouldn't need to be any joining here. I see in your macos implementation it returns the app file, is that necessary? I'm not sure how macos works but ideally that should be a different function or this should be documented so we don't confuse people who use it in the future

File: maxima-lib/src/util/registry.rs [hunk 5]
```diff
@@ -201,25 +205,28 @@ fn register_custom_protocol(protocol: &str, name: &str, executable: &str) -> Res
     let desktop_file_path = format!("{}/.local/share/applications/{}", home, desktop_file_name);
     fs::write(desktop_file_path, desktop_file)?;
 
-    set_mime_type(&format!("x-scheme-handler/{}", protocol), &desktop_file_name)?;
+    set_mime_type(
+        &format!("x-scheme-handler/{}", protocol),
+        &desktop_file_name,
+    )?;
     Ok(())
```
Comment (battledash) [inline maxima-lib/src/util/registry.rs]: Newline pls

File: maxima-cli/Cargo.toml [hunk 6]
```diff
@@ -38,4 +38,4 @@ maxima-resources = { path = "../maxima-resources" }
 
 [[bin]]
 name = "maxima-cli"
-path = "src/main.rs"
\ No newline at end of file
+path = "src/main.rs"
```
Comment (battledash) [inline maxima-cli/Cargo.toml:41]: Add back the newline that was here pls

**Review by battledash — CHANGES_REQUESTED** [review]

File: maxima-bootstrap/Cargo.toml (hunk 1 above)
Comment (Gustash) [reply maxima-bootstrap/Cargo.toml:3]: I don't think it's that big a deal. We can use this for now unless you've got a better description in mind 😅

File: maxima-bootstrap/src/main.rs (hunk 3 above)
Comment (Gustash) [reply maxima-bootstrap/src/main.rs:112]: Our args in macOS are actually just the url to handle, since it's not really coming from the launch args, but from an AppDelegate handler. 

If you notice in `main()`, I'm now sending a slice of the args array skipping the first argument (binary name). This shouldn't be a big deal since we ignored that argument anyway

Comment (battledash) [reply maxima-bootstrap/src/main.rs:112]: Ah yeah I missed the slice up there. Make sure you remove the `args.remove(0);` below that so we don't forget to log an arg in debug

Comment (battledash) [reply maxima-bootstrap/src/main.rs:112]: Actually nvm that doesn't matter, pls ignore me
image

**Review by battledash — APPROVED** [review]

File: maxima-bootstrap/Cargo.toml (hunk 1 above)
Comment (battledash) [reply maxima-bootstrap/Cargo.toml:3]: Yeah it's fine
